feat: rewrite interoperable addresses as a binary representation (#30)
ethereum/L2-interop: specs/addresses/cross-chain-interoperable-addresses-spec.md

Reviewer: What is the mechanism for defining new versions?

Author: defined in actual ERC proposal that future versions merit their own ERC, as they will probably want to define a bunch of extra functionality. Please comment in the [ethereum magicians thread](https://ethereum-magicians.org/t/erc-interoperable-addresses/23365) if you believe this is not clear from the spec or if you believe a different approach (e.g. coordinating versions in CASA?) would be better

# Cross-chain interoperable addresses specification
This document, in its current state, aims to be the starting point for a future address format called 'Interoperable Address' whose aims & properties are fully described in [a separate document](../PROPERTIES.md)

Requires: RFC-4648, ENSIP-11, CAIP-2

# Abstract
An extensible way to describe an address specific to one chain which also includes the information to securely condense it into a human-readable name.

# Out of scope concerns

> Similarly to CAIP-10, this specification is not concerned with the mapping from a chain id to a network name, which might not be surjective (eg: the case where if there are multiple EIP-155 chains with chain id 8453, which one should we call Base?), regarding that resolution a social-layer problem until a future ERC decides to tackle it. Efforts in that front are tracked in [the chain registries document](../docs/chain-registries.md)

Reviewer 2: ```suggestion
Similarly to CAIP-10, this specification does not define the mapping between a chain ID and a network name, as multiple networks can share the same chain ID. For example, if multiple EIP-155 chains have the chain ID 8453, which one should be considered 'Base'? This is treated as a social-layer problem until a future ERC addresses it. Efforts on that front are tracked in [the chain registries document](https://github.com/ethereum/L2-interop/pull/docs/chain-registries.md).
```

Author: I believe semantis are different in your suggestion, with the current one being clearer on that having two networks sharing a chainid being a very impopular edge case

# Definitions
Interoperable Address

> : A binary payload which includes both the information to unambiguously point to a specific address on a specific chain and a _resolver specification_ which, together with this (and potentially future) ERCs, allows conversion to a human-readable name in a way that requires no further trust assumptions than those the wallet software executing said conversion already operates under.

Reviewer 2: Not sure if it's worth mentioning the resolver in this definition, as not all interoperable addresses include resolver information

> : A binary payload which includes both the information to unambiguously point to a specific address on a specific chain and a _resolver specification_ which, together with this (and potentially future) ERCs, allows conversion to a human-readable name in a way that requires no further trust assumptions than those the wallet software executing said conversion already operates under.

Author: that's a fair point, although a bit of social context might clarify the discussion

this standard is being requested by people both trying to get the most succint unambiguous representation of a (chain,address) for intents/message passing purposes and wallet developers/users trying to define how will we finally get to abstract away those pesky 20 bytes of raw pubkey. 

To serve the needs of the former, we make resolver specification optional, but in the near future most people reading this will be in the latter group, considering we plan on re-write ERC-7828 with the contents of this PR

Human-readable name
: A string representation of an interoperable address, condensed into something that is to be consumed by humans.

Target address

> : The (address, chainid) pair a particular Interoperable Address points to.

Reviewer 2: ```suggestion
: The (address, chain ID) pair that a particular Interoperable Address points to."
```

Name-resolving contract

> : A contract where responsibility over converting addresses and/or chainids in a format meant for machine consumption over to parts of a human-readable name.

Reviewer 2: ```suggestion
: A contract responsible for converting addresses and/or chain IDs from a machine-readable format into parts of a human-readable name.
```

# Guarantees provided by the standard

> - Any Interoperable Address is trivially convertible to Interoperable Long Format, `0x8000`, making it a canonical representation for users who need to treat them opaquely.

Reviewer 2: We should replace every occurrence of 'long format' with 'canonical format'

No length restriction is placed on how many more words can an Interoperable Address span. Different versions of the standard can define uses for extra words and the 208 reserved bits.

TODO: consider making this format 31 bytes and define standard ways to pad it so it can be _saved_ in a `bytes` array taking only one word (ABI representation would not be significantly cheaper though).

# Human-readable name format definition

## Syntax
```bnf
<human readable name> ::= <account>@<chain>#<checksum>
<account>             ::= [-:_%a-zA-Z0-9]*
<chain>               ::= [-:_a-zA-Z0-9]*
<checksum>            ::= [0-9A-F]{8}
```

## Rationale
- Chain and account fields' syntax is deliberately chosen to be able to express CAIP-2 namespaces and CAIP-10 account ids, with the caveat that no length restriction is placed, so chains with longer address formats or full 256-bit EVM chainids can be represented.
- Similarly, the account field includes `%` as a valid character to allow for url-encoding of any other characters.

# Interoperable Address version definitions

## `0x8000`: Standard Long Format: arbitrary length addresses, CAIP-10 display format

### Machine-address format

#### First word

> Does not make use of reserved bits

Reviewer 2: ```suggestion
No reserved bits are used
```

```
MSB                                                            LSB
0x8000XXXXXXXX0000000000000000000000000000000000000000000000000000
  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              \--------------- 207-0   Reserved
```

#### Second field

> Serializes the chainid as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays). It is a _field_ and not a _word_, since it may take up more than one word

Reviewer 2: Let's standardize the terminology and consistently use either 'chain ID,' 'chain id,' or 'chainId' in explanations where we mention it
```suggestion
Serializes the chain ID as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays). It is a _field_ and not a _word_, since it may take up more than one word
```

: `0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045#6164da10@eip155:1#618ad0d1`

## `0x0001` : Compact format, ENSIP-11+ERC-3770 name resolution 
The binary representation is identical to the format above, with the exception that the version number is different, indicating the intent for the chain to be displayed using the ERC-3770 shortname and the address to be displayed following mainnet's ENS deployment and ENSIP-11

### Supported chains
Chain must be listed on ethereum-lists/chains on top of the restrictions of `0x0000`.

### Machine-address format
Same as `0x0000`, with `0x0001` version

### Examples:
Machine address
: 
```
MSB                                                            LSB
0x0001618ad0d1000000000001D8DA6BF26964AF9D7EED9E03E53415D37AA96045
  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^
                \------------- 207-160 Chainid
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                            \- 159-0   Address
```

Human-readable name
: `vitalik.eth@eth#618AD0D1`

TODO: example on another chain

# Restrictions for future versions

> - The Interoperable Address MUST include an address on a particular chain and MUST also specify on which chain said address exists, fully defining a target address. A version requiring an indirection layer between data contained in the payload and the two fields described above would be in infringement of this restriction.

Reviewer 2: ```suggestion
- The Interoperable Address MUST include an address and specify the chain it belongs to, fully defining a target address. Any version that adds an indirection layer between the payload data and these fields would violate this requirement.
```

- Interoperable Address versions MAY only be able to represent a subset of the CAIP-2 namespaces.
- Interoperable Address versions MAY define extra fields containing information on how to display the addresses.
- The most significant version bit is reserved to mean an address complies with the Interoperable Long Format, described in Interoperable Address version `0x8000`.
- The second-most significant version bit is reserved to mean the payload includes information on how to display the address to users, but is not normative about how that information is to be encoded. An example of this is provided in Interoperable Address version `0xC000`.

# Requirements for wallet software

> - Wallet software MUST perform all relevant pre-validations, including verifying the checksum, and report any errors to the user. Wallets MAY reject an interoperable address solely on the basis of these checks failing.

Reviewer 2: I would separate the MUST and MAY items into two distinct bullet points.
```suggestion
- Wallet software MUST perform all relevant pre-validations, including verifying the checksum, and report any errors to the user. 
- Wallets MAY reject an interoperable address solely on the basis of these checks failing.
```

# Compatibility with other public-key sharing standards

## W3C DID
TODO

# Appendix A: Binary encoding of CAIP-2 blockchain id
First two bytes are the binary representation of CAIP-2 namespace (see table below), rest are the CAIP-2 reference, whose encoding is namespace-specific and defined below.

## CAIP-2 namespaces' binary representation table

| Namespace | binary key |
| ---       | ---        |
| eip155    | `0x0000`   |
| bip122    | `0x0001`   |

### eip155
The bare `chainid` encoded as a `uint` of the necessary amount of bytes will be used [^1]. This allows for most of existing EVM chainids to fit inside one EVM word.

[^1]: This makes it possible to represent some chains using the full word as their chainid, which CAIP-2 does not support since the set of values representable with 32 `a-zA-Z0-9` characters has less than `type(uint256).max` elements. This is done in an effort to support chains whose id is the output of `keccak256`, as proposed in ERC-7785.

#### Examples

> Ethereum Mainnet: `0x000001` (1, encoded as uint8)

Reviewer 2: For some reason, this is being rendered on the same line in the preview.

Optimism Mainnet: `0x00000A` (10, encoded as uint8)
Ethereum Sepolia: `0x0000aa36a7` (11155111, encoded as uint24)

### bip122
The genesis/fork blockhash is to be stored raw, without encoding/decoding from/to base58btc, and without removing any leading zeroes:

> Note: CAIP-2 has a 32 character limit on chain references which means converting to them will require truncating the reference, and converting _from_ CAIP-2 to this standard is potentially ambiguous (but converting from actual bip122 to this standard will never be)

Reviewer 2: ```suggestion
Note: CAIP-2 limits chain references to 32 characters, so converting to it will require truncation. This makes conversions from CAIP-2 to this standard ambiguous, but not from BIP-122.
```

Author: I believe it's important for a standard to define where it might be ambiguous, and this suggestion hides it. Tried to implement similar improvements to prose though

#### Examples

> Bitcoin Mainnet: `0x0001000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f`

Reviewer 2: These are appearing on the same line in the preview.

Reviewer 2: You can add `-` to avoid this

Author: turned them into dictionary definitions

`0x00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF`:
```
0x
0000000000000000000000000000000000000000000000000000000000000041
00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF
```

`0x00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF0011`:
```
0x
0000000000000000000000000000000000000000000000000000000000000045
00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF
0011000000000000000000000000000000000000000000000000000000000000
```

TODO: 
- Consider tradeoffs of following a head-tail encoding scheme a la ABI, instead of packing the data right after. It would be possible to pack the (relative to current position? absolute to start of payload?) offset into 16 bytes, and use the remaining 16 bytes for `length * 2 + 1`. (pro of this approach: not possible to add extra data between fields which is invisible to parsers like is possible with abi encoding)

# Appendix D: checksum computation
With the intent of:
- Maximizing the difficulty of mining checksum collisions
- Maximize the safety a user gets from knowing the checksum matches

> - Allow maximum flexibility for human-readable names, leveraging the above to make human-readable name collisions permissible

Reviewer 2: NIT: missing the `.`

And considering:
- Binary payloads will often have their own error-correction mechanisms when in transit (eg: if encoded into a QR code, said QR code will have information redundancy & its own checksum ensuring only valid data can be decoded), so that can safely be deemed out of scope.
TODO: Considering the above, does it make sense to store the checksum inside the address?

We propose a way to serialize the 'abstract' components of an address, ignoring the information on how it should be displayed, to achieve the above: 

```solidity

>     function getChecksum(bytes2 caip2Namespace, bytes memory chainid, bytes memory address_)

Reviewer 2: ```suggestion
    function getChecksum(bytes2 caip2Namespace, bytes memory chainId, bytes memory address_)
```

        internal
        view
        returns (bytes4 checksum)
    {

>         bytes memory fullChainid = bytes.concat(caip2Namespace, chainid);

Reviewer 2: ```suggestion
        bytes memory fullChainId = bytes.concat(caip2Namespace, chainId);
```

Human-readable name
: A string representation of an interoperable address, meant to be used by humans.

Target address
: The (address, chain ID) pair a particular Interoperable Address points to.

Name-resolving contract
: A contract responsible for converting addresses and/or chain IDs from a machine-readable format into parts of a human-readable name.

# Guarantees provided by the standard
- Any Interoperable Address is trivially convertible to Interoperable Canonical Format, `0x8000`, making it a canonical representation for users who need to treat them opaquely.
- Checksums are short enough to be compared by humans, and effectively identify a _target address_ independently of any extra data the address may contain.

# Binary Format definition
Word 0 Is the only word which will be common to all versions, and is defined as follows:
```
MSB                                                            LSB
0xXXXXYYYYYYYYZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZZ
  ^^^^------------------------ 255-240 Interoperable Address version

>       ^^^^^^^^ --------------- 239-208 Checksum

Reviewer 3: Checksums are not commonly used in binary format addresses, because it's expected that the communications channel has its own integrity guarantees.

<account>             ::= [-:_%a-zA-Z0-9]*
<chain>               ::= [-:_a-zA-Z0-9]*
<checksum>            ::= [0-9A-F]{8}
```

## Rationale
- Chain and account fields' syntax is deliberately chosen to be able to express CAIP-2 namespaces and CAIP-10 account IDs, with the caveat that no length restriction is placed, so chains with longer address formats or full 256-bit EVM chainids can be represented.
- Similarly, the account field includes `%` as a valid character to allow for url-encoding of any other characters.

# Interoperable Address version definitions

## `0x8000`: Interoperable Canonical Format: arbitrary length addresses, CAIP-10 compatible display format

### Machine-address format

#### First word
No reserved bits are used. They MUST be set to zero, to ensure canonicity of addresses.
```
MSB                                                            LSB
0x8000XXXXXXXX0000000000000000000000000000000000000000000000000000
  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              \--------------- 207-0   Reserved
```

> #### Second field

Reviewer 3: 'Field' should be defined in the definitions section.

Serializes the chain ID as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays). It is a _field_ and not a _word_, since it may take up more than one word.

> #### Third field

Reviewer 3: Any particular reason these two fields are in the opposite order to the text format?

Serializes the address as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays). It is a _field_ and not a _word_, since it may take up more than one word, and may not start at the third word boundary if the second field takes up more than one word.

### Human-readable name resolution
The CAIP-2 namespace is to be rendered alongside the CAIP-10 formatted address and the checksum:

```
<CAIP-10 formatted address>@<CAIP-2 namespace>#<checksum>
```

### Examples

Human-readable name:

> `0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045#6164DA10@eip155:1#618ad0d1`:

Reviewer 3: This is at odds with the rest of the document, which states there's only one checksum for the whole address.

  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              \--------------- 207-0   Not used
```

Second word (second field):
```
MSB                                                            LSB
0x0100000000000000000000000000000000000000000000000000000000000002
  ^^----------------------------------------------------------------- 255-249 bytes array payload
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ---- 9-249   padding
                                                                ^^ -- 0-8    2*1 (payload length)
```

Third word (third field):
```
MSB                                                            LSB
0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045000000000000000000000028
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  ------------------------- 255-97 bytes array payload
                                          ^^^^^^^^^^^^^^^^^^^^^^ ---- 9-96   padding
                                                                ^^ -- 0-8    2*20 (payload length)
```

TODO: example with multi-word fields

> ## `0xC000`: Interoperable Canonical Format With Resolver Info: arbitrary length addresses, arbitrary length name-resolving contract specification

Reviewer 3: I'm not sure I see the usefulness of this - it replaces the non-human-readable address with a human-readable name, but then adds a non-human-readable address that specifies how the name is resolved. The net result is an address that's longer and no easier to read/verify.

### Machine-address format

#### First word
Uses reserved bits as a registry of interfaces the provided resolver conforms to.
```
MSB                                                            LSB
0xC000XXXXXXXX0000000000000000000000000000000000000000000000000000
  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              \--------------- 207-0   Resolver Interface Key
```
The Resolver Interface Key defines how wallets should interact with the smart contract responsible for assigning names to addresses.
Said contract MAY also be responsible for naming the chain on which the target address resides as well as providing a human-readable name for the address, when ERC-7785 or similar standards reach production.

#### Second field

> Serializes the chain ID of the target address  as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays).

Reviewer 3: It doesn't make much sense to me to provide a way of resolving a name to an address, but then still require specifying the chain the address will be for manually - why not require the resolution contract to do that?

Serializes the address of the target address as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays).

#### Fourth field
Serializes the chain ID where the name-resolving contract is located as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays).

#### Fifth field
Serializes the address of the name-resolving contract as described in [Appendix A](#appendix-a-binary-encoding-of-caip-2-blockchain-id) and encodes it into a byte array as described in [Appendix C](#appendix-c-short-encoding-of-byte-arrays).

### Human-readable name resolution
Specific to every Resolver Interface Key

### Resolver Interface Keys

#### `0x0000000000000000000000000000000000000000000000000000`
Responsibility over discovering the interface of the naming smart contract is delegated to the contract itself, via mechanisms comparable to `ERC165`, out of scope for this definition.

#### `0x0000000000000000000000000000000000000000000000000001`

> Naming contract is expected to conform to ENSIP-11, and used to display the address. ENSIP-11 does not name chains, so the CAIP-2 name should be used instead.

Reviewer 3: This seems too vague for an implementer to use. I can't see this being adopted unless it's very concrete about how to resolve an address.

### Examples

Human-readable name:

> `vitalik.eth@eip155:1#618AD0D1`:

Reviewer 3: Where/how is the address of the resolver specified?

- Wallets MAY reject an interoperable address solely on the basis of the checks above failing.

# Encoding considerations
- On-chain actors, such as smart contracts, MUST always receive and return the machine addresses as byte array.
- On-chain actors, such as smart contracts, MAY only accept canonical `0x8000` address versions, with off-chain actors being responsible over converting them to said format.
- Off-chain actors, such as wallets, MAY use the blockchain-native byte array representation, or, where practical MAY alternatively use base64 [^2] as defined in RFC-4648 to encode the former.
- Users SHOULD NOT be shown the base64 or binary representations. Instead, they should be presented with the intended human-readable name or, in cases where that is not possible, wallet software MUST fall back to interpreting it as an Interoperable Address version `0x8000`.

[^2]: Base64 was chosen since it is a more widely implemented standard than base58. Additionally, since machine addresses are not to be directly displayed to the user, the collision-avoidance reasons in favor of base58 do not apply

> # Appendix A: Binary encoding of CAIP-2 blockchain ID

Reviewer: I believe this should be standardized via CAIP. In particular, the binary id table should be CAIP-governed.

Author: Agree. noted this in the eth magicians thread and will try to coordinate with CASA the best way for them to take care of it

## CAIP-2 namespaces' binary representation table

| Namespace | binary key |
| ---       | ---        |
| eip155    | `0x0000`   |
| bip122    | `0x0001`   |

### eip155
The bare `chainid` encoded as a `uint` of the necessary amount of bytes will be used [^1]. This allows for most of existing EVM chain IDs to fit inside one EVM word.

[^1]: This makes it possible to represent some chains using the full word as their chainid, which CAIP-2 does not support since the set of values representable with 32 `a-zA-Z0-9` characters has less than `type(uint256).max` elements. This is done in an effort to support chains whose ID is the output of `keccak256`, as proposed in ERC-7785.

#### Examples
Ethereum Mainnet: `0x000001` (1, encoded as uint8)

Optimism Mainnet: `0x00000A` (10, encoded as uint8)

Ethereum Sepolia: `0x0000aa36a7` (11155111, encoded as uint24)

### bip122
The genesis/fork blockhash is to be stored raw, without encoding/decoding from/to base58btc, and without removing any leading zeroes:

> Note: CAIP-2 limits chain references to 32 characters, so converting to it will require truncating the reference, so converting _from_ CAIP-2 to this standard is potentially ambiguous (but converting from actual bip122 to this standard will never be).

Reviewer: This seems more severe than ambiguity, isn't it? There is no possible conversion from CAIP-2 to this format.

Author: moved the question to the eth-magicians thread as I'm not sure what to do about it either

## eip155
20 bytes of evm addresses trivially stored as the payload.

### Examples
`0xD8DA6BF26964AF9D7EED9E03E53415D37AA96045` -> `0xD8DA6BF26964AF9D7EED9E03E53415D37AA96045`

## solana
base-58 encoded public keys should be decoded and stored as a 32 byte payload

### Examples

`7S3P4HxJpyyigGzodYwHtCxZyUQe9JiBMHyRWXArAaKv` -> `0x5F90554BB3D8C2FC82B6EE59C49AAA143E77F7D49A83E956CE1DBEF17A43F805`

`DYw8jCTfwHNRJhhmFcbXvVDTqWMEVFBX6ZKUmG5CNSKK` -> `0xBA7A74F374AB05B70D114A78112EF0D3F0695A819572C79710B5372000D81AE2`

# Appendix C: Short-encoding of byte arrays
This is a mix between [how bytes and strings are saved into storage](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#bytes-and-string) and how ABI encoding works

This arises out of:
- The need for the shortest possible representation of EVM addresses.
- The absence of a canonical place to delimit the _start_ of the data (in the case of storage strings, this is the keccak256 hash of the storage slot index).

If the data is at most 31 bytes long, the elements are stored in the higher-order bytes (left-aligned) and the lowest-order byte stores the value `length * 2`.

> If the data is 32 bytes or longer, the current word stores `length * 2 + 1`, and data follows in the next words, left-aligned and zero-padded.

Reviewer: What is the reason for padding the data to a multiple of 32?

And are we sure we want to represent the length in 32 bytes? It will be mostly wasted. 32 bits should be more than enough to represent any reasonable length.

Author: thanks for this! removed the requirement to pad lengths to 32 bytes in new version: https://ethereum-magicians.org/t/erc-interoperable-addresses/23365

Similarly to CAIP-10, this specification is not concerned with the mapping from a chain id to a network name, which might not be surjective (eg: the case where if there are multiple EIP-155 chains with chain id 8453, which one should we call Base?), regarding that resolution a social-layer problem until a future ERC decides to tackle it. Efforts in that front are tracked in [the chain registires document](../docs/chain-registries.md)

## Definitions
Interoperable address

> : A binary payload which includes both the information to unambiguosly point to a specific address on a specific chain and a _resolver specification_ which, together with this (and potentially future) ERCs, allows conversion to a human-readable name in a way that requires no further trust assumptions than those the wallet software executing said conversion already operates under.

Reviewer: I don't understand why the mechanism for conversion to human-readable name is a part of the machine address spec. This seems like a completely separate concern.

It also introduces the requirement for canonicalization because e.g. an EVM account can be identified with both 0x0000 and 0x0001 version addresses.

Author: > I don't understand why the mechanism for conversion to human-readable name is a part of the machine address spec

Good question: The idea is to allow an address to instruct the wallet on how to display it, making this standard the common layer upon which users with heterogeneous naming schemes can interoperate (e.g, someone who trusts ethereum-lists/chains with someone who does not, or users of ENS with users of some other competing protocol)

> It also introduces the requirement for canonicalization because e.g. an EVM account can be identified with both 0x0000 and 0x0001 version addresses

yes, the idea is for the checksum to be the tool that enables users to distinguish on whether they are talking about the same address on the same chain or not . Perhaps there's a benefit of having a fully canonical representation of an address I'm not seeing?

Also, one of the properties we want for the various formats is to always be able to extract the (chain,address) pair cheaply from the payload and be able to not care about all other potential fields with minimal overhead. We would be opposed to an interoperable address version which stores a punycode string in the payload and requires an ENS lookup to obtain the actual address, for example.

Reviewer: My interest is using these identifiers on-chain. I want an identifier that I can treat opaquely, i.e. that I don't have to inspect and manipulate in order to use safely. This requires that there be a single way to identify each account.

For example, if I receive a message from a sender the bridge claims to be `A` I want to be able to index into a mapping `messages[A]` and store/fetch all past messages of `A` (this is just an example). If the sender can be identified by an alternative address `A'`, since `messages[A] != messages[A']`, I cannot do this safely without assuming that the bridge will always consistently use the same identifier format, and that the bridge will never be swapped out for another that uses a different format.

As far as I understand, the proposal in this PR doesn't achieve this, because there are alternative identifiers with different formats.

I don't see how the checksum addresses this concern since on its own it doesn't have enough security due to its small size.

> The idea is to allow an address to instruct the wallet on how to display it, making this standard the common layer upon which users with heterogeneous naming schemes can interoperate (e.g, someone who trusts ethereum-lists/chains with someone who does not, or users of ENS with users of some other competing protocol)

My first intuition is that the wallet should define its own trust assumptions. I don't see why it would make sense for it to accept what the address requests, if it's not already trusted by the wallet. And even if that was the goal, I don't think it should be part of the machine address because it's a presentation concern.

Reviewer: > Our idea is to enshrine one of the representation (`0x8000`, in the current state of the spec) as the 'canonical' one, and guarantee all other representations can be casted to it, so you'd be able to do that with a gas overhead similar to abi-decoding a struct

Unless the set of formats is closed, this is not possible. A contract needs to manually implement decoding for every format, so when a new format is defined the contract is no longer able to canonicalize all addresses.

Reviewer 4: This is ok, since, compact formats are aimed for inter-ecosystem communication, e.g. op-superchain messaging, where all contracts and actors involved know how to encode/decode this special sub-type by leveraging shared assumptions on the system they are running on.

Reviewer: Ok, so you're suggesting something like ERC-7786 should always operate on an address of type `0x8000`, right? For example, if a relayer submits a message for verification any addresses in another format would cause a revert.

Reviewer: It would be nice if the canonical format was `0x0000`. It obviously doesn't matter but I think it sends a clearer signal.

Author: > Ok, so you're suggesting something like ERC-7786 should always operate on an address of type 0x8000. For example, if a relayer submits a message for verification any addresses in another format would cause a revert.

Added this to the standard's text :sparkles: . (i.e. the _right_ for contracts to revert on addresses which are not of that type) I'd like to also standardize the error message/kind but given that this might be used on non-evm runtimes idk how feasible would that be.

> It would be nice if the canonical format was 0x0000.

While that would be a nice modification right now, I believe setting the first bit to one to express the address is of this Standard Long Format type makes it more futureproof since future address versions can be of this type and also include some other info. `0xC000` is a current example of that which mostly services wallets, but it would be possible to define a version also including relevant info for a _payment request_, for example

                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                            \- 159-0   Address
```

`Interoperable Address version`
: Always `0x0000`

`Checksum`
: Computed as described in [Appendix D](#Appendix-D)

`Chainid`
: `uint48` containing the chainId for the target chain

`Address`
: Native EVM address

#### Human-readable name resolution
The CAIP-2 namespace is to be rendered alongside the `0x`-prefixed EIP-55 address and the checksum

```
<address>@<CAIP-2 namespace>#<checksum>
```

#### Examples:
Machine address
: 
```
MSB                                                            LSB
0x0000618ad0d1000000000001D8DA6BF26964AF9D7EED9E03E53415D37AA96045
  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^
                \------------- 207-160 Chainid
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                            \- 159-0   Address
```

Human-readable name
: `0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045#6164da10@eip155:1#618ad0d1`

### `0x0001` : Compact format, ENSIP-11+ERC-3770 name resolution 

> The binary representation is identical to the format above, with the exception that the version number is different, indicating the intent for the chain to be displayed using the ERC-3770 shortname and the address to be displayed following mainnet's ENS deployment and ENSIP-11

Reviewer: I don't think display concerns should not be part of the machine address format.

> The binary representation is identical to the format above, with the exception that the version number is different, indicating the intent for the chain to be displayed using the ERC-3770 shortname and the address to be displayed following mainnet's ENS deployment and ENSIP-11

Reviewer: I don't think display concerns should be part of the machine address format.

## Definitions
Interoperable address
: A binary payload which includes both the information to unambiguosly point to a specific address on a specific chain and a _resolver specification_ which, together with this (and potentially future) ERCs, allows conversion to a human-readable name in a way that requires no further trust assumptions than those the wallet software executing said conversion already operates under.

Human-readable name
: A string representation of an interoperable address, condensed into something that is to be consumed by humans.

## Machine address format definition #1: easiest-calldata

In order to maximize ease of use with most existing infrastructure, the format is deliberately solidity-ABI-compatible. For environments where solidity-ABI is not a native encoding, userspace parsers could be easily written.

This standard defines the format at the binary level, but for better readability, solidity structs with solidity types will be used throughout this document.

The foundation for the Interoperable Address format is the following struct:

```solidity
struct InteroperableAddress {
    bytes32 version_checksum_reserved;
    bytes chainId;
    bytes address_;

> }

Author: yes, at the cost of having 2 extra words only storing the length for each (plus 2 more words storing the offset of the value from the start of the payload, a cost of ABI compatibility)

```

where:

`version`
: Defines the Interoperable Address version, which will be useful to know if to expect any extra fields at runtime. `uint16` stored in the most significant bits (240-255) of the `version_checksum_reserved`

`checksum`
: A 4-byte value computed over the entirety of the Interoperable Address payload (details below). `bytes4` stored in bits (239-208) of `version_checksum_reserved`

`reserved`
: Bytes that ABI-encoding would otherwise pad to zero, which can have specific uses in future versions of the standard.

`chainid`
: Binary representation of the CAIP-2 blockchain id. See [Appendix A](#Appendix-A)

`address_`

> : Raw binary representation of the address in the chain. For `eip155` chains, this is trivial. Other chains can have different types of addresses which include more than just a public key, and those will be serialized as described in [Appendix B](#Appendix-B).

Reviewer 2: ```suggestion
: Raw binary representation of the address in the chain. For EIP-155 chains, this is trivial. Other chains can have different types of addresses which include more than just a public key, and those will be serialized as described in [Appendix B](#Appendix-B).
```

Author: I deliberately chose to use the same string as in the CAIP-2 namespace, which will also show up in the binary representation table.

### Word-by-word definition of the format

#### Word 0:
```
MSB                                                            LSB
0x0000000000000000000000000000000000000000000000000000000000000000
  ^^^^------------------------ 255-240 Interoperable Address version
      ^^^^^^^^ --------------- 239-208 Checksum
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                \------------- 207-0   Reserved
```

#### Field 1:
Binary-encoded CAIP-2 blockchain id (following [Appendix A](#Appendix-A)) inside a short-packed bytes array, as described in [Appendix C](#Appendix-C), taking 1 or more 32-byte words.

#### Field 2:
Binary-encoded address (following [Appendix B](#Appendix-B)) inside a short-packed bytes array, as described in [Appendix C](#Appendix-C), taking 1 or more 32-byte words.

#### Field 3+:
Optionally used by the fields defined above, in the case chain ids or addresses are long enough to merit it.

> Further fields may be used by future versions of the standard

Reviewer 2: ```suggestion
Additional fields may be specified in future versions of the standard
```

## Checksum definition & validation
The high-level description of how to compute the checksum is:
- Set checksum to zero.
- Serialize the machine address.
- Compute the `keccak256` hash of the serialized address.

> - Insert the first four bytes of the hash where the checksum is supposed to be

Reviewer 2: ```suggestion
- Insert the first four bytes of the hash where the checksum should be.
```

- Serialize the machine address again

TODO: should the checksum go over the entire address, including the instructions on how to display it, or only the fields required for version zero? I'm specifying the former but leaning for the latter

### Example

> Assuming machine address format #1, CAIP-2 eip155 namespace has binary id `0x0001`, network is eth mainnet, and address in that network `0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045`:

Reviewer 2: ```suggestion
Assuming machine address format #1, CAIP-2 EIP-155 namespace has binary ID of `0x0001`, the network is Ethereum mainnet, and the address on that network is `0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045`:
```

```
InteroperableAddress memory addy = InteroperableAddress({
    version_checksum_reserved: bytes32(uint256(uint16(2))) << 240,
    chainId: hex'000101',
    address_: hex'd8dA6BF26964aF9D7eEd9e03E53415D37aA96045'
});
```

> Serlialization with checksum set to zero:

Reviewer 2: ```suggestion
Serialization with checksum set to zero:
```

``` 
0x
0000000000000000000000000000000000000000000000000000000000000020
0002000000000000000000000000000000000000000000000000000000000000
0000000000000000000000000000000000000000000000000000000000000060
00000000000000000000000000000000000000000000000000000000000000a0
0000000000000000000000000000000000000000000000000000000000000003
0001010000000000000000000000000000000000000000000000000000000000
0000000000000000000000000000000000000000000000000000000000000014
d8da6bf26964af9d7eed9e03e53415d37aa96045000000000000000000000000
```
keccak256 of the above:
```
0xeb43550c612da0271a1f59bf6e9e4f3572139769ba88e9d05337d31fa1c2d2c2
```
Inserting the checksum into the address:
```
addy.version_checksum_reserved |= bytes6(keccak256(abi.encode(addy))) >> 16;
```

> Serialized address including checksum:

Reviewer 2: ```suggestion
Serialized address including the checksum:
```

## Rationale
- In order to guarantee future resolvers don't share prefixes with existing resolvers in an ambiguous way, and therefore human-readable names are always mappable to a single machine address, this standard defines the human-readable names to start with the resolver version. (TODO: ideally we should be smarter about this and resolve it in a way that doesn't expose this implementation quirk to users)

## Encoding considerations

> - on-chain actors, such as smart contracts, MUST always receive and return the machine addresses as byte array

Reviewer 2: ```suggestion
- On-chain actors, such as smart contracts, MUST always receive and return the machine addresses as byte arrays.
```

> - off-chain actors, such as wallets, MAY use the blockchain-native byte array representation, or, where practical MAY alternatively use base64 as defined in RFC-4648 to encode the former.

Reviewer 2: ```suggestion
- Off-chain actors, such as wallets, MAY use the blockchain-native byte array representation, or, where practical MAY alternatively use base64 as defined in RFC-4648 to encode the former.
```

> - users SHOULD NOT be shown the base64 or binary representations, instead, they should be shown the intended human-readable name or, in cases where that is not possible, falling back to interpreting it as Interoperable Address version zero.

Reviewer 2: ```suggestion
- Users SHOULD NOT be shown the base64 or binary representations. Instead, they should be presented with the intended human-readable name or, if that is not possible, they should fall back to interpreting it as an Interoperable Address version zero.
```

TODO: probably coupled to ERC-7785, currently out of scope

## Rationale
- Checksum algorithm is independent of used resolution method to allow wallets to validate an interoperable address' checksum despite not being able to generate its human-readable name
- Ideally we'd want the resolution method to be fully abstracted away from the user, but that might not be achievable in every case. The next best thing is to prefix every human-readable name with a string denoting which method to use, which prevents the user ever being prompted to choose different addresses based on the resolution method used for each, which would be a more confusing and riskier UX 

> - Base64 was chosen since it is a more widely implemented standard than base58, and since machine addresses are not to be directly displayed to the user, the collision-avoidance reasons in favor of base58 do not apply

Reviewer 2: ```suggestion
- Base64 was chosen since it is a more widely implemented standard than base58. Additionally, since machine addresses are not meant to be directly displayed to users, the collision-avoidance advantages of base58 do not apply
```

## Security considerations

## Appendix A: Binary encoding of CAIP-2 blockchain id

> First two bytes are the binary representation of CAIP-2 namespace (see table below), rest are the CAIP-2 reference.

Reviewer 2: ```suggestion
The first two bytes are the binary representation of CAIP-2 namespace (see table below), while the remaining bytes correspond to the CAIP-2 reference.
```

> In the special case of eip155 chains, the bare `chainid` encoded as a `uint` of the necessary amount of bytes will be used [^1]. This allows for the existing EVM chainids fit inside one EVM word.

Reviewer 2: ```suggestion
For EIP-155 chains, the `chainid` is encoded as a `uint` using the necessary number of bytes [^1]. This ensures that existing EVM chain IDs fit within a single EVM word.
```

> Reference will be the ASCII-encoded value from CAIP-2 (from, and not including, the colon, up to the end of the string) in all other cases.

Reviewer 2: ```suggestion
In all other cases, the reference will be the ASCII-encoded value from CAIP-2, starting after the colon and extending to the end of the string.
```

> [^1]: This makes it possible to represent some chains using the full word as their chainid, which CAIP-2 does not support since the maximum value representable with 32 `a-zA-Z0-9` characters is less than `type(uint256).max`. This is done in an effort to support chains whose id is the output of `keccak256`, as proposed in ERC-7785.

Reviewer 2: ```suggestion
[^1]: This allows certain chains to use the full word as their chainid, which CAIP-2 does not support since the maximum value representable with 32 `a-zA-Z0-9` characters is less than `type(uint256).max`. This intends to support chains whose ID is derived from `keccak256`, as proposed in ERC-7785.
```

Author: semantics are different for 'allows certain chains to' vs 'makes it possible to represent some chains'. We cannot tell chains what to do :sweat_smile: , and are instead trying to make whatever chains choose to do representable in this standard

### CAIP-2 namespaces' binary representation table
TODO

### Examples
TODO

## Appendix B: Binary encoding of addresses
TODO, will be similar to ENSIP-9 and/or CAIP-50

### Examples
TODO

## Appendix C: Short-encoding of byte arrays
This is a mix between [how bytes and strings are saved into storage](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#bytes-and-string) and how ABI encoding works

This arises out of:

> - The need to have the shortest possible representation for EVM addresses.

Reviewer 2: ```suggestion
- The need for the shortest possible representation of EVM addresses.
```

> - Not having a canonical place to delimit the _start_ of the data (in the case of storage strings, it's the keccak256 hash of the storage slot index).

Reviewer 2: ```suggestion
- The absence of a canonical delimiter for the _start_ of the data (in the case of storage strings, this is the `keccak256` hash of the storage slot index).
```

> If the data is at most 31 bytes long, the elements are stored in the higher-order bytes (left aligned) and the lowest-order byte stores the value `length * 2`.

Reviewer 2: ```suggestion
If the data is at most 31 bytes long, the elements are stored in the higher-order bytes (left-aligned) and the lowest-order byte stores the value `length * 2`.
```
